openedx/core/djangoapps/user_api/accounts/api.py

```diff
@@ -429,32 +584,189 @@ def _validate_username(username):
         AccountUsernameInvalid
 
     """
+    try:
+        _validate_unicode(username)
+    except UnicodeError as invalid_unicode_err:
+        raise AccountUsernameInvalid(str(invalid_unicode_err))
+    _validate_username_type(username)
+    _validate_username_of_good_length(username)
+
+    try:
+        validate_slug(username)
+    except ValidationError:
+        raise AccountUsernameInvalid(
+            u"Username '{username}' must contain only A-Z, a-z, 0-9, -, or _ characters".format(username=username)
+        )
+
+
+def _validate_username_type(username):
+    """Validate the username instance type.
+
+    :param username: The proposed username (unicode).
+    :return: None
+    :raises: AccountUsernameInvalid
+    """
     if not isinstance(username, basestring):
         raise AccountUsernameInvalid(u"Username must be a string")
 
+
+def _validate_username_of_good_length(username):
+    """Validate that the username is less than or equal to
+    the maximum username length allowed, and greater than or
+    equal to the minimum username length allowed.
+
+    :param username: The proposed username (unicode).
+    :return: None
+    :raises: AccountUsernameInvalid
+    """
+    _validate_username_satisfies_min_length(username)
+    _validate_username_satisfies_max_length(username)
```

Comment:
I disagree. At this point, we've got six different functions to handle checking the length of a string. Perhaps consolidate to one function that takes a string, a max length, and a min length, then call it from each of `validate_username_of_good_length`, `_validate_email_of_good_length` and `_validate_password_of_good_length`? Then, we'd have small, cohesive blocks of logic. Each `_validate` function could pull the relevant settings and pass them in, the single length-checking function can make its check, and the calling function can output the relevant error message if needed.